### PR TITLE
Faster Qulacs expval

### PR DESCRIPTION
**Pull request by thisac**
Use native expectation value calculations.

Seemingly works now. There were issues with rotations, so I just used `self._pre_rotated_states` instead, and also made it into a copy of `self._state`, rather than a reference, which was the case before.

Two questions left:

1. Do we need rotations at all now, and if not, why should the `QulacsDevice.state` property return the post-rotated states (otherwise the tests seem to fail)? It was actually using the `self._pre_rotated_states` before, but since that was just a reference to `self._state`, it was actually using the post-rotated states all along.

2. Not sure what to do about the Hadamard and Hermitian observables. Currently it uses the old expval calculations in those specific cases.

### CODE REVIEW

**Comment by josh146** [comment]
Is it not working because the circuit rotations have to be turned off?

Just a guess, I haven't looked at the source code!

File: pennylane_qulacs/qulacs_device.py [hunk 1]
```diff
@@ -295,12 +295,44 @@ def analytic_probability(self, wires=None):
...
+        }
+
+        if self.analytic:
+            qulacs_observable = Observable(self.num_wires)
+            if isinstance(observable.name, list):
+                observables = [observable_map[obs] for obs in observable.name]
+            else:
+                observables = [observable_map[observable.name]]
+
+            if None not in observables:
```
Comment (mariaschuld) [inline pennylane_qulacs/qulacs_device.py:315]: Just to understand, could one not use the native methods for all observables that are not None here?

**Review by mariaschuld — APPROVED** [review]
Looks good, if the test pass. But it would be really nice to have a quick benchmark between this branch and master before merging. Just to see if native observables make a huge difference. 

I'm not sure, but could one just run the benchmark suite with this device?

File: pennylane_qulacs/qulacs_device.py (hunk 1 above)
Comment (thisac) [reply pennylane_qulacs/qulacs_device.py:315]: I'm actually not sure how that would be done to be honest. This was simply the easiest way to implement it, and I think should, in most cases, be fine.

**Comment by thisac** [comment]
Just ran the benchmarking suite with some of the benchmarks comparing the new exval implementation with how it was done before, and the results are very close. Not sure if these benchmarks are the best at testing this though. Any thoughts @josh146 @mariaschuld?

```
Benchmarking PennyLane 0.12.0-dev
10 loops, 5 runs

Timing: 'Entangling layers' benchmark on qulacs.simulator, QNode
Timing per loop (new): [0.1559389  0.14596113 0.1462956  0.14792621 0.14919876]
Timing per loop (old): [0.14912802 0.14583779 0.1512493  0.14579161 0.14617127]

Timing: 'Jacobian evaluation uniform' benchmark on qulacs.simulator, QNode
Timing per loop (new): [0.17294835 0.18056541 0.18726398 0.18580912 0.17770734]
Timing per loop (old): [0.18384072 0.18606518 0.18081585 0.18195038 0.18745033]

Timing: 'Jacobian evaluation end weighted' benchmark on qulacs.simulator, QNode
Timing per loop (new): [0.18155389 0.18747421 0.17934829 0.18798632 0.1830495 ]
Timing per loop (old): [0.18422577 0.1826246  0.1842201  0.19189538 0.18537905]

Timing: 'IQP circuit' benchmark on qulacs.simulator, QNode
Timing per loop (new): [0.0040996  0.00362196 0.0036716  0.00364643 0.00439197]
Timing per loop (old): [0.00431251 0.00338209 0.00335678 0.00351857 0.00347142]

Timing: 'Nearest neighbour' benchmark on qulacs.simulator, QNode
Timing per loop (new): [0.04634738 0.04701537 0.04720423 0.0467239  0.04659007]
Timing per loop (old): [0.04789591 0.04222938 0.04152561 0.04139612 0.04126724]
```

**Comment by mariaschuld** [comment]
> Just ran the benchmarking suite with some of the benchmarks comparing the new exval implementation with how it was done before, and the results are very close. Not sure if these benchmarks are the best at testing this though. Any thoughts @josh146 @mariaschuld?
> 
> ```
> Benchmarking PennyLane 0.12.0-dev
> 10 loops, 5 runs
> 
> Timing: 'Entangling layers' benchmark on qulacs.simulator, QNode
> Timing per loop (new): [0.1559389  0.14596113 0.1462956  0.14792621 0.14919876]
> Timing per loop (old): [0.14912802 0.14583779 0.1512493  0.14579161 0.14617127]
> 
> Timing: 'Jacobian evaluation uniform' benchmark on qulacs.simulator, QNode
> Timing per loop (new): [0.17294835 0.18056541 0.18726398 0.18580912 0.17770734]
> Timing per loop (old): [0.18384072 0.18606518 0.18081585 0.18195038 0.18745033]
> 
> Timing: 'Jacobian evaluation end weighted' benchmark on qulacs.simulator, QNode
> Timing per loop (new): [0.18155389 0.18747421 0.17934829 0.18798632 0.1830495 ]
> Timing per loop (old): [0.18422577 0.1826246  0.1842201  0.19189538 0.18537905]
> 
> Timing: 'IQP circuit' benchmark on qulacs.simulator, QNode
> Timing per loop (new): [0.0040996  0.00362196 0.0036716  0.00364643 0.00439197]
> Timing per loop (old): [0.00431251 0.00338209 0.00335678 0.00351857 0.00347142]
> 
> Timing: 'Nearest neighbour' benchmark on qulacs.simulator, QNode
> Timing per loop (new): [0.04634738 0.04701537 0.04720423 0.0467239  0.04659007]
> Timing per loop (old): [0.04789591 0.04222938 0.04152561 0.04139612 0.04126724]
> ```

@thisac Well, let's think a bit more, which circuit would expose the changes on this branch? One where PL would natively make a decomposition of the measurement observable but qulacs expectation values can now directly evaluate it?

**Comment by mariaschuld** [comment]
@thisac please tag me if another review is needed :)

**Comment by thisac** [comment]
@mariaschuld Ready for a (final) review. I added a test for the `expval()` function just checking that the native expectation values are used with the supported observables. Other than that, it's the same as before.

I tested running the code below and got:
* `734 µs ± 10.4 µs per loop (mean ± std. dev. of 200 runs, 1000 loops each)` when using PennyLane's `expval()`
* `432 µs ± 17.6 µs per loop (mean ± std. dev. of 200 runs, 1000 loops each)` when using Qulacs' native `expval()`

```
import pennylane as qml
import numpy as np

dev = qml.device("qulacs.simulator", wires=2)

@qml.qnode(dev)
def circuit():
    return qml.expval(qml.PauliX(wires=0) @ qml.PauliY(wires=1))

%timeit -r 200 circuit()
```

Not sure how representative this test is of real-world cases though, but at least it's faster here (and should hopefully not be slower in any cases). :slightly_smiling_face:

File: pennylane_qulacs/qulacs_device.py [hunk 2]
```diff
@@ -295,12 +295,44 @@ def analytic_probability(self, wires=None):
...
 
+    def expval(self, observable):
+        observable_map = {
+            "PauliX": "X",
+            "PauliY": "Y",
+            "PauliZ": "Z",
+            "Identity": "I",
+            "Hadamard": None,
+            "Hermitian": None,
+        }
```
Comment (mariaschuld) [inline pennylane_qulacs/qulacs_device.py:306]: Should we not put this in the header to make it global? Or is there a reason to redefine it in every expval call?

**Review by mariaschuld — APPROVED** [review]
Great, thanks @thisac!

Only one minor comment to move the observable map?

Otherwise ready for merge as far as I can judge.

**Pull request merged**